File: src/main/java/org/example/test_maksik/service/NoteService.java
package org.example.test_maksik.service;

import lombok.RequiredArgsConstructor;
import org.example.test_maksik.dto.NoteRequestDto;
import org.example.test_maksik.dto.NoteResponseDto;
import org.example.test_maksik.entity.NoteEntity;
import org.example.test_maksik.exception.NotFoundException;
import org.example.test_maksik.repository.NoteRepository;
import org.springframework.data.domain.Page;
import org.springframework.stereotype.Service;

import org.springframework.data.domain.Pageable;

import java.time.LocalDateTime;
import java.util.List;

@Service
@RequiredArgsConstructor
public class NoteService {
    private final NoteRepository repository;
    private NoteResponseDto toDto(NoteEntity note){
        return NoteResponseDto.builder().id(note.getId()).title(note.getTitle()).content(note.getContent())
                .createdAt(note.getCreatedAt()).updatedAt(note.getUpdatedAt()).build();
    }

    public Page<NoteResponseDto> getAll(Pageable pageable){
        return repository.findAllByDeletedAtIsNull(pageable).map(this::toDto);
    }

    public NoteResponseDto getById(Long id){
        NoteEntity note = repository.findByIdAndDeletedAtIsNull(id).orElseThrow(() -> new NotFoundException("Note not foun d"));
        return toDto(note);
    }

    public NoteResponseDto create(NoteRequestDto dto){
        NoteEntity note = NoteEntity.builder().title(dto.getTitle()).content(dto.getContent())
                .createdAt(LocalDateTime.now()).build();
        return toDto(repository.save(note));
    }

    public NoteResponseDto update(Long id, NoteRequestDto dto){
        NoteEntity note = repository.findByIdAndDeletedAtIsNull(id).orElseThrow(() -> new NotFoundException("Note not foun d"));
        note.setTitle(dto.getTitle());
        note.setContent(dto.getContent());
        note.setUpdatedAt(LocalDateTime.now());
        return toDto(repository.save(note));
    }

    public void delete(Long id){
        NoteEntity note = repository.findByIdAndDeletedAtIsNull(id).orElseThrow(() -> new NotFoundException("Note not foun d"));
        note.setDeletedAt(LocalDateTime.now());
        repository.save(note);
    }

    public List<NoteResponseDto> searchByTitle(String title){
        return repository.findByTitleContainingIgnoreCaseAndDeletedAtIsNull(title).stream().map(this::toDto).toList();
    }

    public List<NoteResponseDto> searchByContent(String content){
        return repository.findByContentContainingIgnoreCaseAndDeletedAtIsNull(content).stream().map(this::toDto).toList();
    }
}

File: src/main/kotlin/com/example/demo/services/NoteService.kt
package com.example.demo.services

import com.example.demo.dtos.NoteDto
import com.example.demo.dtos.NoteResponseDto
import com.example.demo.entities.NoteEntity
import com.example.demo.exceptions.NoteNotFoundException
import com.example.demo.repositories.NoteRepository
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.stereotype.Service

@Service
class NoteService(private val repository: NoteRepository) {

    fun getAll(pageable: Pageable): Page<NoteResponseDto> {
        return repository.findAllByDeletedAtIsNull(pageable).map { it.toResponseDto() }
    }

    fun getById(id: Long): NoteResponseDto {
        val note = repository.findByIdAndDeletedAtIsNull(id) ?: throw NoteNotFoundException(id)
        return note.toResponseDto()
    }

    fun searchByTitle(title: String): List<NoteResponseDto> {
        return repository.findAllByTitleContainingIgnoreCaseAndDeletedAtIsNull(title).map { it.toResponseDto() }
    }

    fun searchByContent(content: String): List<NoteResponseDto> {
        return repository.findAllByContentContainingIgnoreCaseAndDeletedAtIsNull(content).map { it.toResponseDto() }
    }

    fun create(dto: NoteDto): NoteResponseDto {
        val note = NoteEntity(title = dto.title, content = dto.content)
        return repository.save(note).toResponseDto()
    }

    fun update(id: Long, dto: NoteDto): NoteResponseDto {
        val note = repository.findByIdAndDeletedAtIsNull(id) ?: throw NoteNotFoundException(id)
        note.title = dto.title
        note.content = dto.content
        note.updatedAt = System.currentTimeMillis()
        return repository.save(note).toResponseDto()
    }

    fun delete(id: Long) {
        val note = repository.findByIdAndDeletedAtIsNull(id) ?: throw NoteNotFoundException(id)
        note.deletedAt = System.currentTimeMillis()
        repository.save(note)
    }

    private fun NoteEntity.toResponseDto() = NoteResponseDto(
        id = id,
        title = title,
        content = content,
        createdAt = createdAt,
        updatedAt = updatedAt
    )
}
